### ISSUE TITLE
How to get session.dirty rows in created order

### ISSUE THREAD

**Issue by valhuber** [sqlalchemy/sqlalchemy#5603]
**Describe your question**
I know session.dirty returns an identity set, and these are unordered.  But I see no property of the list items I could sort on to process, for example, orders before order-details.

**Example - please use the [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) guidelines if possible**

**Complete stack trace, if applicable**

**Versions**
 - OS: MacOS
 - Python: 3.8.3
 - SQLAlchemy: 1.3.19
 - Database: sqlite
<img width="1402" alt="unordered dirty" src="https://user-images.githubusercontent.com/5579631/93730184-58a5d180-fb7c-11ea-85d3-a3085bc1f1ae.png">

 - DBAPI: session.dirty

**Additional context**

**Useful links**
- [Get Support](https://www.sqlalchemy.org/support.html) on the website
- The [documentation](https://docs.sqlalchemy.org/en/latest/) website
- The [UsageRecipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) wiki
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sqlalchemy) tag
- SQLAlchemy [Google group](http://groups.google.com/group/sqlalchemy)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment]
"order before order details" doesn't sound as much like "created" order as it does "dependency".   I could create an OrderDetail object with None for "order" and then set it later to a new Order, right?     also the "create" case would be in session.new, not session.dirty.

if you want to sort things in the order that events happened, you again can use the event model to track new object constructions as well as mutations to those objects and sort it all in terms of the session they belong to, e.g. InstanceEvents and AttributeEvents: https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=instanceevents#sqlalchemy.orm.events.InstanceEvents  https://docs.sqlalchemy.org/en/13/orm/events.html?#attribute-events

but if you want to actually sort in order of foreign key dependency then you'd need to get a hold of the toplogical dependency graph that the ORM creates for this, which is based explicitly on the graph of mapper() objects as nodes linked to each other using relationship() as edges.    I'd have to work that up for you how to do that.

so basically i need a more specific definition of "created order".

**Comment by valhuber** [comment]
I was looking for “the order things happen”, and was hoping that the sqlalchemy engine could provide some sort of “touched_number” automatically, set when the first touch is made to a given row.  I presume it would be using those same events, but that the engine could save me the work.  It seems it would be generally useful.

Is this practical?

I had thought of the dependency order, but that sounds complex and worth avoiding.

Thanks for the quick reply.

**Comment by zzzeek** [comment]
this ordering is set up as the flush plan and occurs when flush actually occurs.    I could show you how to create this flush plan ahead of time but this is not really public API.    you can certainly intercept most SQL statements within flush as they happen though.   it's not clear if you want this before the whole flush actually happens or as it happens.  for as-it-happens, very easy use the mapper before_insert / before_update / before_delete events.  https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=mapperevents#sqlalchemy.orm.events.MapperEvents

**Comment by zzzeek** [comment]
hi Val looks like you've worked this out so closing this for now.
